[25.0 backport] libnet: Don't forward to upstream resolvers on internal nw

vvoland opened the pull request:
- backport: https://github.com/moby/moby/pull/47538

**- What I did**

Picked up Albin's change from https://github.com/moby/moby/pull/46609 - and added a regression test.

Commit cbc2a71c2 makes `connect` syscall fail fast when a container is only attached to an internal network. Thanks to that, if such a container tries to resolve an "external" domain, the embedded resolver returns an error immediately instead of waiting for a timeout.

This commit makes sure the embedded resolver doesn't even try to forward to upstream servers - this is important when the host's DNS server is running on a localhost address. In this case, the internal resolver's upstream DNS requests are made from the host's network namespace, so they work even when the network is declared as 'internal'. Communication with external DNS servers is unexpected for an internal network.

**- How Albin did it**

Add a way to enable/disable upstream forwarding to the embedded resolver. When an endpoint joins/leaves a sandbox, this forwarding policy is modified based on whether there's an endpoint providing external connectivity.

**- How to verify it**

New regression test.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
Do not forward requests to external DNS servers for a container that is only connected to an 'internal' network. Previously, requests were forwarded if the host's DNS server was running on a localhost address, like systemd's 127.0.0.53.
```

**- A picture of a cute animal (not mandatory but encouraged)**

akerouanton committed:
libnet: Don't forward to upstream resolvers on internal nw

Commit cbc2a71c2 makes `connect` syscall fail fast when a container is
only attached to an internal network. Thanks to that, if such a
container tries to resolve an "external" domain, the embedded resolver
returns an error immediately instead of waiting for a timeout.

This commit makes sure the embedded resolver doesn't even try to forward
to upstream servers.

Co-authored-by: Albin Kerouanton <albinker@gmail.com>
Signed-off-by: Rob Murray <rob.murray@docker.com>
(cherry picked from commit 790c3039d0ca5ed86ecd099b4b571496607628bc)
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Note, this didn't cherry-pick cleanly due to https://github.com/moby/moby/pull/47041 not being backported.
The conflict was only in `integration/networking/resolvconf_test.go` - so I just removed the `TestResolvConfLocalhostIPv6` test which was added in that PR.

@robmry PTAL

robmry commented:
> Note, this didn't cherry-pick cleanly due to #47041 not being backported. The conflict was only in `integration/networking/resolvconf_test.go` - so I just removed the `TestResolvConfLocalhostIPv6` test which was added in that PR.

Thank you ... I think the test is probably failing because it relies on env var DOCKER_TEST_RESOLV_CONF_PATH to point the daemon at the test's own resolv.conf instead of the host's - and that env var was also added by the resolvconf refactoring.

We could either just not-backport the test (and test it manually), or backport the [container_operations.go](https://github.com/moby/moby/pull/47041/files#diff-6e3774954611ecbf441ac0aa5841f11926c1e618b391ff64c330cd64e8875756) change from that PR?

robmry committed:
Environment variable to override resolv.conf path.

If env var DOCKER_TEST_RESOLV_CONF_PATH is set, treat it as an override
for the 'resolv.conf' path.

Added as part of resolv.conf refactoring, but needed by back-ported test
TestInternalNetworkDNS.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
> ..., or backport the [container_operations.go](https://github.com/moby/moby/pull/47041/files#diff-6e3774954611ecbf441ac0aa5841f11926c1e618b391ff64c330cd64e8875756) change from that PR?

I did that.

vvoland commented:
Thanks! "LGTM" 😄

robmry reviewed (approved):
LGTM too ... but I guess that's cheating! @akerouanton, could you take a look?

akerouanton reviewed (commented):
LGTM
